psi4/src/psi4/libmints/local.cc

```diff
@@ -549,4 +550,544 @@ void PMLocalizer::localize() {
+            std::vector<double> Qs;
+            for (int A = 0; A < natom; A++) {
+                Qs.push_back(std::fabs(Qp[A][i]));
+            }
+            std::sort(Qs.begin(), Qs.end(), std::greater<double>());
+            double Qtot = 0.0;
+            for (int A = 0; A < natom && A < 2; A++) {
+                Qtot += Qs[A];
+            }
+            if (Qtot < stars_completeness_) {
+                pi_orbs.push_back(i);
+            }
+        }
+
+        std::vector<int> centers;
+        for (int i2 = 0; i2 < pi_orbs.size(); i2++) {
+            int i = pi_orbs[i2];
+            int ind = 0;
+            for (int A = 0; A < natom; A++) {
+                if (std::fabs(Qp[A][i]) >= std::fabs(Qp[ind][i])) {
+                    ind = A;
+                }
+            }
+            centers.push_back(ind);
+        }
+        std::sort(centers.begin(), centers.end());
+
+        outfile->Printf("    *** Stars Analysis ***\n\n");
+        outfile->Printf("    Pi Centers: ");
+        for (int ind = 0; ind < centers.size(); ind++) {
+            outfile->Printf("%3d ", centers[ind] + 1);
+        }
+        outfile->Printf("\n\n");
+    }
+
+    std::shared_ptr<Matrix> Focc2 = linalg::triplet(U, Focc, U, true, false, false);
+    std::shared_ptr<Matrix> U2 = IBOLocalizer::reorder_orbitals(Focc2, ranges);
+
+    std::shared_ptr<Matrix> Uocc3 = linalg::doublet(U, U2, false, false);
+    std::shared_ptr<Matrix> Focc3 = linalg::triplet(Uocc3, Focc, Uocc3, true, false, false);
+    std::shared_ptr<Matrix> Locc3 = linalg::doublet(Cocc, Uocc3, false, false);
+
+    L = linalg::doublet(U2, L, true, false);
+    std::shared_ptr<Matrix> Q = orbital_charges(L);
+
+    std::map<std::string, std::shared_ptr<Matrix> > ret;
+    ret["L"] = Locc3;
+    ret["U"] = Uocc3;
+    ret["F"] = Focc3;
+    ret["Q"] = Q;
+
+    ret["L"]->set_name("L");
+    ret["U"]->set_name("U");
+    ret["F"]->set_name("F");
+    ret["Q"]->set_name("Q");
+
+    L_ = ret["L"];
+    U_ = ret["U"];
+
+    return ret;
+}
+
+void IBOLocalizer::localize() {
+    throw PSIEXCEPTION("Non-parameter input function localize() in IBOLocalizer is not defined. Use localize(Cocc, Focc, ranges2).");
+}
+
+std::shared_ptr<Matrix> IBOLocalizer::orbital_charges(std::shared_ptr<Matrix> L) {
+    double** Lp = L->pointer();
+    int nocc = L->rowspi()[0];
+    int nmin = L->colspi()[0];
+    int natom = true_atoms_.size();
+
+    auto Q = std::make_shared<Matrix>("Q", natom, nocc);
+    double** Qp = Q->pointer();
+
+    for (int i = 0; i < nocc; i++) {
+        for (int m = 0; m < nmin; m++) {
+            Qp[iaos_to_atoms_[m]][i] += Lp[i][m] * Lp[i][m];
+        }
+    }
+
+    return Q;
+}
+
+void IBOLocalizer::print_charges(double scale) {
+    if (!A_) build_iaos();
+
+    std::shared_ptr<Matrix> L = linalg::triplet(C_, S_, A_, true, false, false);
+
+    int nocc = L->rowspi()[0];
+    int natom = true_atoms_.size();
+
+    std::shared_ptr<Matrix> Q = orbital_charges(L);
```

Comment:
There is no reason for `Q`, `N`, or `A` to be a pointer. Change these to be raw objects. That'll mean changing the signature of `orbital_charges`.

Reply:
I think Q (orbital_charges) should remain a `shared_ptr`, since the fsapt code needs to access the variable, and we don't wanna copy matrices.

Reply:
I don't see how a copy would occur.

For expressions like `variable_type my_variable = my_function(my_arguments)`, C++ is smart enough not to create the return value and then copy it to the variable. It's called _return value optimization_.

Reply:
I changed N to be a raw object, but not Q, since the fisapt code expects Q to be a pointer, and not a raw object.